sqlglot/optimizer/scope.py

```diff
@@ -272,6 +272,20 @@ def columns(self):
                 ancestor = column.find_ancestor(
                     exp.Select, exp.Qualify, exp.Order, exp.Having, exp.Hint, exp.Table, exp.Star
                 )
+
+                is_except_col = False
+                if isinstance(ancestor, exp.Star):
+                    except_cols = ancestor.args.get("except") or []
+                    replace_cols = ancestor.args.get("replace") or []
+                    cols = set(
+                        replace_col
+                        for col in replace_cols
+                        for replace_col in col.find_all(exp.Column)
+                    )
+                    # Exclude the columns in EXCEPT() from the scope unless they're used in REPLACE, i.e in
+                    # "SELECT * EXCEPT (col1) REPLACE (COALESCE(1, col1) AS col2)"", col1 must be kept in scope
+                    is_except_col = column in except_cols and column not in cols
```

Comment:
Is there a reason to exclude the `EXCEPT` columns from `scope.columns`, even if they're not used in a `REPLACE`?

Reply:
To match the semantics of these engines I think they should be removed from scope (e.g. you won't have stray columns in the `scope.columns` which could be a risk for `validate_qualify_columns`), and this was also added recently by us so I tried to work around it.

I don't have an example right now to prove this and I also didn't experiment with removing it so it may turn out fine, do you have any hunch for it?

Reply:
Not sure how we match the semantics of these engines by not including the `EXCLUDE` columns in `scope.columns`, can you elaborate? I'd add them because that feels more intuitive to me & because it seems it'd simplify the logic. Also, it's weird that this clause stands out for no good reason, but I'm not sure about edge cases.

Wanna try this and see if it breaks stuff?

Reply:
Take the following query in Snowflake:

```SQL
WITH x AS (SELECT 1 AS a) SELECT * EXCLUDE(a) RENAME(a AS d) FROM x;
```

This throws the error _`Column 'A' does not exist`_, so I think the equivalent logic in SQLGlot is that `a` is removed from `scope`, right?

However, I'm on board with removing the extra logic if it's not needed, will attempt it now

Reply:
> so I think the equivalent logic in SQLGlot is that a is removed from scope, right?

Not necessarily, because statically you know there's a column `a` within `EXCLUDE`. This particular example throws an error because you exclude it and then try to rename the excluded column, but if you chose a different column in the `RENAME` clause it would work just fine.

Reply:
I tried to remove the logic in `scope.py` but a few tests broke, and I'm not sure if they're fixed easily or without replicating the same logic at different places. Two examples of what's incorrectly produced without it (before vs after `qualify_columns()`):

1. The columns inside `EXCEPT` are qualified, which doesn't seem to work in either SF or BQ:
```SQL
WITH x AS (SELECT 1 AS a) SELECT * EXCEPT(a) FROM x;
WITH x AS (SELECT 1 AS a) SELECT * EXCEPT(x.a) FROM x AS x;
```

2. Both instances of `a` here will be transformed to `COALESCE(x.a, y.a)` when in reality we'd expect only 1 instance:
```SQL
SELECT a, * EXCEPT (a) FROM x AS x LEFT JOIN x AS y USING (a);
SELECT
 COALESCE(x.a, y.a) AS a,
 COALESCE(x.a, y.a) AS a,
 x.b AS b,
 y.b AS b 
 FROM x AS x LEFT JOIN x AS y ON x.a = y.a;
```

```

 The common denominator here afaict is that we don't want to collect the columns inside of `EXCEPT(...)` because the rest of the qualification logic will find their instances in that list and will proceed with transforming/qualifying them when in reality they should be removed.